torch/csrc/dynamo/guards.cpp

```diff
@@ -1449,6 +1450,80 @@ class NO_TENSOR_ALIASING : public RelationalGuard {
   ska::flat_hash_map<PyObject*, std::nullptr_t> _unique_tensors;
 };
 
+/**
+ * Symbolic Shape Guard.
+ */
+class SYMBOLIC_SHAPE_GUARD : public RelationalGuard {
+ public:
+  SYMBOLIC_SHAPE_GUARD(
+      py::int_ nargs,
+      py::int_ py_addr,
+      py::object py_addr_keep_alive,
+      py::object verbose_code_parts)
+      : RelationalGuard(std::move(verbose_code_parts)),
+        _py_addr_keep_alive(std::move(py_addr_keep_alive)),
+        _args_seen{0} {
+    _nargs = PyLong_AsSize_t(nargs.ptr());
+    if (PyErr_Occurred()) {
+      throw py::value_error(
+          "SYMBOLIC_SHAPE_GUARD expected a non-negative number of arguments.");
+    }
+    uintptr_t addr = PyLong_AsUnsignedLongLong(py_addr.ptr());
+    if (PyErr_Occurred()) {
+      throw py::value_error(
+          "SYMBOLIC_SHAPE_GUARD expected an address to a C function.");
+    }
+    _guard_check_fn = reinterpret_cast<int64_t (*)(int64_t*)>(addr);
+    _args = std::vector<int64_t>(_nargs);
+  }
+
+  bool check_nopybind(PyObject* value) override {
+    // We know that these arguments came from IndexedSource(TensorPropertyGuard)
+    // and therefore no need to check that the value is a Tuple[int, int].
+    PyObject* py_idx = PyTuple_GET_ITEM(value, 0);
+    PyObject* py_val = PyTuple_GET_ITEM(value, 1);
+    Py_ssize_t iarg = PyLong_AsSsize_t(py_idx);
+    _args[iarg] = PyLong_AsLongLong(py_val);
+    _args_seen++;
+
+    if (_args_seen == _nargs) {
+      _args_seen = 0;
+      return _guard_check_fn(_args.data());
+    } else {
+      // We don't have all the values yet. Return true until we get all.
+      return true;
+    }
+  }
+
+  GuardDebugInfo check_verbose_nopybind(PyObject* value) override {
+    if (!PyTuple_Check(value)) {
+      return GuardDebugInfo(false, "Non tuple found!", 0);
+    } else if (PyTuple_Size(value) != 2) {
+      return GuardDebugInfo(false, "Tuple of size not 2 found!", 0);
+    } else if (!PyLong_Check(PyTuple_GET_ITEM(value, 0))) {
+      return GuardDebugInfo(false, "Non integer found!", 0);
+    } else if (!PyLong_Check(PyTuple_GET_ITEM(value, 1))) {
+      return GuardDebugInfo(false, "Non integer found!", 0);
+    }
+    bool result = check_nopybind(value);
+
+    if (!result) {
+      return GuardDebugInfo(false, verbose_code_parts(), 0);
+    }
+    return GuardDebugInfo(true, 1);
+  }
+
+  void reset_state() final {
+    _args_seen = 0;
```

Comment:
Do you have to clear _args?

Reply:
We always have a `_nargs`-sized vector so that we could write values in the correct place. Clearing it would make it a zero-sized vector. We would have to overwrite them with zeros or some default value which is unnecessary.